tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -409,35 +410,53 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// \brief Expands input glob into a collection of full paths to files.
+/// \brief Recursively expand the glob to take care of potential wildcard
+/// specials for subdirectories in the glob.
+/// \param[in] l The list of full paths to files.
 /// \param[in] glob The glob to expand.
-/// \throws std::runtime_error If the directory part of the glob refers to a
+/// \throws std::runtime_error If the directory parts of the glob refer to a
 ///         path that cannot be opened.
-/// \return A vector of strings, the fully expanded paths to the files referred
-///         to by the glob.
 ///
-/// The two parts of the glob (directory, fileglob) are separated. The directory
-/// is first expanded via TSystem::ExpandPathName then opened via
-/// TSystem::OpenDirectory. If the directory can be opened, then the remaining
-/// fileglob is used as regex expression (via TRegexp) and the function stores
-/// those files in the directory that match the regex.
-std::vector<std::string> ExpandGlob(const std::string &glob)
+/// If the glob contains a wildcard special for subdirectories, the three parts
+/// of the glob (directory, subdirectoryglob, remainder) are separated.
+/// Otherwise the glob is expanded to (directory, fileglob).
+/// The directory is first expanded via TSystem::ExpandPathName then opened via
+/// TSystem::OpenDirectory. If the directory can be opened, then current
+/// glob is used as regex expression (via TRegexp) to find subdirectories or
+/// store those files in the directory that match the regex.
+void RecursiveGlob(TList &l, const std::string &glob)
 {
-   // TODO: Also implement this pattern in some other free function
-   // This is practically Python's os.path.split
    std::string dirname;
-   std::string basename;
-
-   const auto slashpos = glob.rfind('/');
-
-   if (slashpos != std::string::npos) {
-      // Separate the glob into its two components
-      dirname = glob.substr(0, slashpos);
-      basename = glob.substr(slashpos + 1);
+   std::string basename; // current glob to expand, could be a directory or file.
+   std::string remainder;
+
+   // This list of characters is currently only defined inside TString::MaybeWildcard() at
+   // https://github.com/root-project/root/blob/5df0ef8bfa3c127e554e845cd6582bc0b4d7f96a/core/base/src/TString.cxx#L960.
+   const char *wildcardSpecials = "[]*?";
+
+   const auto wildcardPos = glob.find_first_of(wildcardSpecials);
+   // Get the closest slash, to the left of the first wildcard
+   auto slashLPos = glob.rfind("/", wildcardPos);
+   // Get the closest slash, to the right of the first wildcard
+   const auto slashRPos = glob.find("/", wildcardPos);
```

Comment:
```suggestion
   auto slashLPos = glob.rfind('/', wildcardPos);
   // Get the closest slash, to the right of the first wildcard
   const auto slashRPos = glob.find('/', wildcardPos);
```
If you're only looking for single characters, the overload that takes only single characters is preferred because it has less overhead